### ISSUE TITLE
Gradient of jax arrays produces numpy arrays

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#20620]
### Description

In some cases, when the gradient is zero, `grad` produces a NumPy array having dtype float0.  This breaks common usage patterns with the Array API, and makes it harder to write platform-agnostic code.

Would it make sense to reconsider https://github.com/google/jax/issues/4433 and have float0 be a Jax dtype?  Or should users of `grad` convert outputs to jax arrays if necessary?

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.4
python: 3.11.8 (main, Feb 22 2024, 17:25:49) [GCC 11.4.0]

**Comment by NeilGirdhar** [comment]
For me, the easiest solution was the third solution (adjusting my calls to `get_namespace`):

```python
def fix_zero(x: Any) -> Any:
    if x.dtype == float0:
        return jnp.zeros(x.shape)
    return x

def get_namespace_fixed(self, *x):
        values = [fix_zero(getattr(self, field.name))
                  for field in fields(self)] + list(x)
        return get_namespace(*values)
```

Does it really make sense to consider an array of dtype `float0` to be a numpy array?  The Numpy namespace doesn't even know anything about float0.  It seems to me that an array with a given library's dtypes should be considered to belong to that library.

**Comment by NeilGirdhar** [comment]
That sounds good.  I just hope that whatever representation you choose derives from `jax.Array`, and is recognized by `get_namespace` as a Jax array :smile:

**Comment by patrick-kidger** [comment]
I will note that I get pretty scared by the idea of changing `float0`.

I've had to add careful handling for this in various places, so I worry that changing this -- and breaking the existing code that handles it -- will lead to difficult-to-debug errors for any downstream users. It's such an edge case.

**Comment by NeilGirdhar** [comment]
@patrick-kidger Are you using the Array API yet?

**Comment by NeilGirdhar** [comment]
FYI: [I modified array-API-compat](https://github.com/data-apis/array-api-compat/pull/130) to work around this problem for now, but it would still be better to smooth out the quirks on the Jax side when there's time.
